src/Kestrel.Transport.Sockets/Internal/SocketAwaitable.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Net.Sockets;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal
+{
+    public class SocketAwaitable : ICriticalNotifyCompletion
+    {
+        private static readonly Action _callbackCompleted = () => { };
+
+        private Action _callback;
+        private int _bytesTransfered;
+        private SocketError _error;
+
+        public SocketAwaitable GetAwaiter() => this;
+        public bool IsCompleted => _callback == _callbackCompleted;
+
+        public int GetResult()
+        {
```

Comment:
If this is called when the task hasn't been completed, what happens? You get the data from the last result? Seems sketchy even for internal code. If a check on `_callback` is going to impact perf, maybe an Assert?

Reply:
Though I see that the corefx version linked below doesn't do that check either. I don't know, maybe I don't care. Just seems safer to slap an assert in there or something.

Reply:
This class makes no attempt to validate that it's being called/used correctly. This is the same as LibuvAwaitable and AwaitableSocketAsyncEventArgs as you point out.

Would a `Debug.Assert(_callback == _callbackCompleted)` work for you? I guess we could do the same in LibuvAwaitable for consistency.

Reply:
It would certainly make me feel a little better. But I get it, it's an internal component and we use it "right" so it's not a big thing.